### PR TITLE
Report income expence

### PR DESCRIPTION
**Pull request by lomamech**
Adding income and expense report  …
- Adding Income Expense Service
- Creation of configuration interface allowing the configuration or Reports
- Adding interface for showing the report
- Adding routes and API for Income and Expense Reports

Pending Html Report

@sfount

### CODE REVIEW

File: client/src/js/services/IncomeExpenseService.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
```
Comment (mbayopanda) [inline client/src/js/services/IncomeExpenseService.js]: You can remove the `use strict`, it is injected automatically

File: client/src/js/services/IncomeExpenseService.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * @name CashflowService
+ * @description
+ * The `IncomeExpenseService` component provide a mean to get IncomeExpense data by requesting
+ * to the /finance/incomeExpense API
+ */
```
Comment (mbayopanda) [inline client/src/js/services/IncomeExpenseService.js:6]: 👍 The description of the service is clear. The `IncomeExpenseService` is not a component, could you update the description?

File: client/src/partials/finance/incomeExpense/config.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+'use strict';
+
```
Comment (mbayopanda) [inline client/src/partials/finance/incomeExpense/config.js]: The `use strict` is not necessary at the client side code, `gulp` add it automatically.

File: client/src/partials/finance/incomeExpense/config.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+// dependencies injection
+IncomeExpenseConfigController.$inject = [ '$state', '$translate', 'CashboxService', 'NotifyService' ];
+
+/**
+ * IncomeExpense config controller
+ *
+ * @description
+ * This controller is responsible of cash flow report, that report include
+ * all incomes minus all depenses
```
Comment (mbayopanda) [inline client/src/partials/finance/incomeExpense/config.js]: The report is for `cashflow` ?

File: client/src/partials/finance/incomeExpense/index.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+// dependencies injection
+IncomeExpenseController.$inject = [ '$state', '$stateParams', 'ModalService', 'LanguageService', 'DateService'];
+
+/**
+ * IncomeExpense controller
+ *
+ * @description
+ * This controller is responsible of cash flow report, that report include
+ * all incomes minus all depenses
```
Comment (mbayopanda) [inline client/src/partials/finance/incomeExpense/index.js]: Could you fix the description

File: client/src/partials/finance/incomeExpense/report.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
+'use strict';
+
```
Comment (mbayopanda) [inline client/src/partials/finance/incomeExpense/report.js]: Not necessary

File: client/src/partials/finance/incomeExpense/report.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  'util',
+];
+
+/**
+ * IncomeExpense report controller
+ *
+ * @description
+ * This controller is responsible of Income Expense report, that report include
+ * all incomes minus all depenses
+ */
```
Comment (mbayopanda) [inline client/src/partials/finance/incomeExpense/report.js]: 👍

File: server/controllers/finance/reports/incomeExpense/index.js [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        posting_journal.currency_id, posting_journal.trans_id, posting_journal.description, posting_journal.comment, posting_journal.origin_id, posting_journal.user_id 
+      FROM posting_journal
+        WHERE posting_journal.account_id=? AND (posting_journal.trans_date >=? AND posting_journal.trans_date <=?)
+      ) UNION (
+      SELECT general_ledger.project_id, general_ledger.uuid, general_ledger.trans_date, general_ledger.debit_equiv, general_ledger.credit_equiv,
+        general_ledger.debit, general_ledger.credit, general_ledger.account_id, general_ledger.record_uuid, general_ledger.entity_uuid,  general_ledger.reference_uuid,
+        general_ledger.currency_id, general_ledger.trans_id, general_ledger.description, general_ledger.comment, general_ledger.origin_id, general_ledger.user_id 
+      FROM general_ledger
+        WHERE general_ledger.account_id=? AND (general_ledger.trans_date >=? AND general_ledger.trans_date <=?)
+      )
```
Comment (mbayopanda) [inline server/controllers/finance/reports/incomeExpense/index.js:115]: As you suppose it, we must be careful when comparing dates without using MySQL DATE function.

File: client/src/js/services/IncomeExpenseService.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+/**
+ * @name CashflowService
```
Comment (jniles) [inline client/src/js/services/IncomeExpenseService.js]: This also is the IncomeExpenseService, not a CashflowService.

**Review by jniles — COMMENTED** [review]
I'm handing this review over to @sfount, since I'm not familiar with the report configuration on the client.

@sfount, can you take a look and make sure this is up to standards?

**Comment by sfount** [comment]
@jniles This was implemented before the archived report page (https://github.com/IMA-WorldHealth/bhima-2.X/pull/832) was committed. 

This pull request should be brought up to those standards; that can either be through a merge and branch or by forking this branch directly.

**Comment by jniles** [comment]
Okay, thanks.  I was a bit lost as to best practices.

I can take a look at #832, fork this branch, and make sure it is up to date with those changes.  That should round of Income Expense while Chris is working on Client/Debtor reports.

**Comment by jniles** [comment]
@lomamech, don't worry about these comments, I'll make the changes as I work on updating this PR.

**Comment by jniles** [comment]
@lomamech, I think I know what the errors are.  I'll try to merge this today.

**Pull request closed without merging**